src/mds/MDSRank.cc

```diff
@@ -3311,6 +3313,36 @@ void MDSRank::command_dump_inode(Formatter *f, const cmdmap_t &cmdmap, std::ostr
   }
 }
 
+void MDSRank::command_dump_dir(Formatter *f, const cmdmap_t &cmdmap, std::ostream &ss)
+{
+  std::lock_guard l(mds_lock);
+  std::string path;
+  bool got = cmd_getval(cmdmap, "path", path);
+  if (!got) {
+    ss << "missing path argument";
+    return;
+  }
+
+  bool dentry_dump = false;
+  cmd_getval(cmdmap, "dentry_dump", dentry_dump);
+
+  CInode *in = mdcache->cache_traverse(filepath(path.c_str()));
+  if (!in) {
+    ss << "directory inode not in cache";
+    return;
+  }
+
+  f->open_array_section("dirs");
+  frag_vec_t leaves;
+  in->dirfragtree.get_leaves_under(frag_t(), leaves);
+  for (const auto& leaf : leaves) {
+    CDir *dir = in->get_dirfrag(leaf);
+    if (dir)
+      mdcache->dump_dir(f, dir, dentry_dump);
```

Comment:
If the dirfrag is not in cache then perhaps put something in the JSON indicating so.

Reply:
Makes sense. Let's do this as a follow up for this change since the current revision is tested and would probably get merged soon.

Reply:
> If the dirfrag is not in cache then perhaps put something in the JSON indicating so.

@batrick 
Just to confirm, did you mean add the status of [this](https://github.com/ceph/ceph/pull/49756/files#diff-f91675b3e095c571fa2e401547ac27a8f0b1575788f0df3321a8f20a5949f90dR3331) to the JSON or check the `dirfrag` especially ?

Reply:
If the dirfrag is not loaded in the mds cache (the `else` case), we should include that info in the JSON.

Reply:
@vshankar 
I've checked this and I see this https://github.com/ceph/ceph/blob/main/src/mds/CDir.h#L122 is getting printed in the json already. So looks like if the value `complete` is included in the `states` array in json, then it's in the cache. Otherwise `complete` won't be there.

Reply:
> I've checked this and I see this https://github.com/ceph/ceph/blob/main/src/mds/CDir.h#L122 is getting printed in the json already. So looks like if the value `complete` is included in the `states` array in json, then it's in the cache. Otherwise `complete` won't be there.

I understand that. The ask here is if a dirfrag is not yet loaded in the MDS cache, then  the get_dirfrag() call will return NULL (see CInode::get_dirfrag).

Reply:
When get_dirfrag() call returns NULL, dir is NULL, so we can't call `mdcache->dump_dir()`. So create a new json or print something in the else case?